### PR TITLE
Scrapper, Margarita Velmisova - 22FPL1

### PR DESCRIPTION
**Pull request by velgarita**

### CODE REVIEW

**Comment by demid5111** [comment]
nothing to review

File: lab_5_scrapper/scrapper.py [hunk 1]
```diff
@@ -168,6 +347,14 @@ def _fill_article_with_text(self, article_soup: BeautifulSoup) -> None:
...
             article_soup (bs4.BeautifulSoup): BeautifulSoup instance
         """
+        main_div = article_soup.find('div', class_='article__content clearfix')
+        texts = []
+        if main_div:
+            all_ps = main_div.find_all('p', class_='article__text')
+            for p in all_ps:
+                texts.append(p.text)
+
+        self.article.text = ''.join(texts)
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py:357]: 🎖️ very nice

File: lab_5_scrapper/scrapper.py [hunk 2]
```diff
@@ -127,11 +268,43 @@ def _extract_url(self, article_bs: BeautifulSoup) -> str:
...
+                continue
+
+            article_soup = BeautifulSoup(response.text, features='html.parser')
+            new_url = self._extract_url(article_soup)
+
+
+            while new_url:
+                if len(self.urls) == self.config.get_num_articles():
+                    break
+                self.urls.append(new_url)
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py:302]: do you check for duplicates?

File: lab_5_scrapper/scrapper.py [hunk 3]
```diff
@@ -127,11 +268,43 @@ def _extract_url(self, article_bs: BeautifulSoup) -> str:
...
         Find articles.
         """
+        seed_urls = self.get_search_urls()
+
+        for seed_url in seed_urls:
+            response = make_request(seed_url, self.config)
+            if not response.ok:
+                continue
+
+            article_soup = BeautifulSoup(response.text, features='html.parser')
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py:295]: use lxml. for 8+ mark it is required

File: lab_5_scrapper/scrapper.py [hunk 4]
```diff
@@ -127,11 +268,43 @@ def _extract_url(self, article_bs: BeautifulSoup) -> str:
...
+            new_url = self._extract_url(article_soup)
+
+
+            while new_url:
+                if len(self.urls) == self.config.get_num_articles():
+                    break
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py:301]: infinite loop if not enough articles on the page?

File: lab_5_scrapper/scrapper.py [hunk 5]
```diff
@@ -187,6 +391,10 @@ def unify_date_format(self, date_str: str) -> datetime.datetime:
         Returns:
             datetime.datetime: Datetime object
         """
+        dt_object = datetime.datetime.strptime(date_str, '%d.%m.%Y %H:%M')
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py:394]: return immediately

File: lab_5_scrapper/scrapper.py [hunk 6]
```diff
@@ -100,6 +229,14 @@ def make_request(url: str, config: Config) -> requests.models.Response:
     Returns:
         requests.models.Response: A response from a request
     """
+    sleep(randrange(3))
+
+    headers = config.get_headers()
+    timeout = config.get_timeout()
+
+    response = requests.get(url=url, headers=headers, timeout=timeout)
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py:237]: return immediately, stop making intermediate variables

File: lab_5_scrapper/scrapper.py [hunk 7]
```diff
@@ -26,11 +95,64 @@ def _extract_config_content(self) -> ConfigDTO:
...
+        ):
+            raise IncorrectSeedURLError
+
+        if not (
+            isinstance(config.total_articles, int)
+            and config.total_articles > 0
+        ) or isinstance(config.total_articles, bool):
+            raise IncorrectNumberOfArticlesError
+
+        if not (
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py:130]: open brackets, avoid conditions with negation, if possible

File: lab_5_scrapper/scrapper.py [hunk 8]
```diff
@@ -26,11 +95,64 @@ def _extract_config_content(self) -> ConfigDTO:
...
 
     def _validate_config_content(self) -> None:
         """
         Ensure configuration parameters are not corrupt.
         """
+        config = self._extract_config_content()
+
+        if not(
+                isinstance(config.seed_urls, list)
+                and all(re.match('https?://(www.)?', url) for url in config.seed_urls)
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py:120]: more strict checks are needed - that these are seed urls for selected besite

File: lab_5_scrapper/scrapper.py [hunk 9]
```diff
@@ -26,11 +95,64 @@ def _extract_config_content(self) -> ConfigDTO:
         Returns:
             ConfigDTO: Config values
         """
+        with open(self.path_to_config, 'r', encoding='utf-8') as file:
+            config_content = json.load(file)
+        return ConfigDTO(
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py:100]: re-work in kwargs-style notation

File: lab_5_scrapper/scrapper.py [hunk 10]
```diff
@@ -2,9 +2,65 @@
...
+
+class IncorrectSeedURLError(Exception):
+    """
+    Seed URL does not match standard pattern
+    """
+
+
+class NumberOfArticlesOutOfRangeError(Exception):
+    """
+    Total number of articles is out of range
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py:31]: put period at the end of each docstring

File: lab_5_scrapper/scrapper.py [hunk 11]
```diff
@@ -2,9 +2,65 @@
...
+
+class IncorrectTimeoutError(Exception):
+    """
+    Timeout value must be a positive integer less than 60
+    """
+
+
+class IncorrectVerifyError(Exception):
+    """
+    Verify certificate and Headless mode values must boolean
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py:61]: must BE boolean

**Review by demid5111 — CHANGES_REQUESTED** [review]
blazingly fast progress. maybe you decide to start recusrive crawler?

File: lab_5_scrapper/scrapper.py [hunk 12]
```diff
@@ -2,9 +2,65 @@
 Crawler implementation.
```
[the remark this answers is not in the document]
Comment (velgarita) [reply lab_5_scrapper/scrapper.py:2]: Thank you! Next time I will definitely do it locally

File: requirements.txt [hunk 13]
```diff
@@ -1 +1,10 @@
+beautifulsoup4==4.12.3
+datetime==5.5
+pytest-shutil==1.7.0
+python-path==0.1.3
+requests==2.31.0
+stanza==1.8.1
+
+
+
```
[the remark this answers is not in the document]
Comment (velgarita) [reply requirements.txt:9]: Removed it

File: requirements.txt [hunk 14]
```diff
@@ -1 +1,10 @@
+beautifulsoup4==4.12.3
+datetime==5.5
+pytest-shutil==1.7.0
+python-path==0.1.3
+requests==2.31.0
+stanza==1.8.1
```
[the remark this answers is not in the document]
Comment (velgarita) [reply requirements.txt:6]: Removed it

File: requirements.txt [hunk 15]
```diff
@@ -1 +1,10 @@
+beautifulsoup4==4.12.3
+datetime==5.5
+pytest-shutil==1.7.0
```
[the remark this answers is not in the document]
Comment (velgarita) [reply requirements.txt:3]: Ok, I mixed it up with built in shutil module. I removed it

File: requirements.txt [hunk 16]
```diff
@@ -1 +1,10 @@
+beautifulsoup4==4.12.3
+datetime==5.5
```
[the remark this answers is not in the document]
Comment (velgarita) [reply requirements.txt:2]: Fixed it

File: lab_5_scrapper/scrapper.py (hunk 7 above)
Comment (velgarita) [reply lab_5_scrapper/scrapper.py:130]: Changed it

File: lab_5_scrapper/scrapper.py (hunk 8 above)
Comment (velgarita) [reply lab_5_scrapper/scrapper.py:120]: Changed it

File: lab_5_scrapper/scrapper.py (hunk 9 above)
Comment (velgarita) [reply lab_5_scrapper/scrapper.py:100]: Done

File: lab_5_scrapper/scrapper.py (hunk 10 above)
Comment (velgarita) [reply lab_5_scrapper/scrapper.py:31]: Done

File: lab_5_scrapper/scrapper.py (hunk 11 above)
Comment (velgarita) [reply lab_5_scrapper/scrapper.py:61]: Fixed it

File: lab_5_scrapper/scrapper.py (hunk 6 above)
Comment (velgarita) [reply lab_5_scrapper/scrapper.py:237]: Fixed it here and where was possible

File: lab_5_scrapper/scrapper.py (hunk 3 above)
Comment (velgarita) [reply lab_5_scrapper/scrapper.py:295]: Changed it

File: lab_5_scrapper/scrapper.py (hunk 4 above)
Comment (velgarita) [reply lab_5_scrapper/scrapper.py:301]: Changed it

File: lab_5_scrapper/scrapper.py (hunk 2 above)
Comment (velgarita) [reply lab_5_scrapper/scrapper.py:302]: I check it in '_extract_url' method, but I added check here too

File: lab_5_scrapper/scrapper.py (hunk 1 above)
Comment (velgarita) [reply lab_5_scrapper/scrapper.py:357]: Thank you! 🥹

File: lab_5_scrapper/scrapper.py (hunk 5 above)
Comment (velgarita) [reply lab_5_scrapper/scrapper.py:394]: Fixed it

File: lab_5_scrapper/scrapper.py [hunk 17]
```diff
@@ -204,13 +420,42 @@ def prepare_environment(base_path: Union[pathlib.Path, str]) -> None:
     Args:
         base_path (Union[pathlib.Path, str]): Path where articles stores
     """
+    if not base_path.is_dir():
+        base_path.mkdir(parents=True, exist_ok=True)
+
+    if any(base_path.iterdir()):
+        shutil.rmtree(base_path)
+        base_path.mkdir(parents=True, exist_ok=True)
```
[the remark this answers is not in the document]
Comment (velgarita) [reply lab_5_scrapper/scrapper.py:428]: Changed it

File: lab_5_scrapper/scrapper.py [hunk 18]
```diff
@@ -204,13 +420,42 @@ def prepare_environment(base_path: Union[pathlib.Path, str]) -> None:
...
+        article_id = index + 1
+        html_parser = HTMLParser(full_url=url, article_id=article_id, config=configuration)
+        article = html_parser.parse()
+
+        if isinstance(article, Article):
+            to_raw(article)
+            to_meta(article)
+
+    print('Ready!')
+    print(len(crawler.urls))
```
[the remark this answers is not in the document]
Comment (velgarita) [reply lab_5_scrapper/scrapper.py:457]: Removed it

File: lab_5_scrapper/scrapper.py [hunk 19]
```diff
@@ -204,13 +420,42 @@ def prepare_environment(base_path: Union[pathlib.Path, str]) -> None:
...
+
 
 if __name__ == "__main__":
-    main()
+    main()
```
[the remark this answers is not in the document]
Comment (velgarita) [reply lab_5_scrapper/scrapper.py:461]: Fixed it

File: requirements.txt [hunk 20]
```diff
@@ -1 +1,10 @@
...
+pytest-shutil==1.7.0
+python-path==0.1.3
```
[the remark this answers is not in the document]
Comment (velgarita) [reply requirements.txt:3]: Removed it

File: lab_5_scrapper/scrapper.py (hunk 9 above)
Comment (demid5111) [reply lab_5_scrapper/scrapper.py:100]: now we can shorten it to dict unpacking

File: requirements.txt [hunk 21]
```diff
@@ -1 +1,3 @@
-
+beautifulsoup4==4.12.3
+pytest-shutil==1.7.0
+requests==2.31.0
```
Comment (demid5111) [inline requirements.txt:3]: newline

File: requirements.txt (hunk 15 above)
[the remark this answers is not in the document]
Comment (demid5111) [reply requirements.txt:2]: you did not

File: lab_5_scrapper/scrapper.py [hunk 22]
```diff
@@ -196,6 +400,14 @@ def parse(self) -> Union[Article, bool, list]:
             Union[Article, bool, list]: Article instance
         """
 
+        response = make_request(self.full_url, self.config)
+        if response.ok:
+            article_bs = BeautifulSoup(response.text, features='html.parser')
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py:405]: for 8 and 10 you have to use lxml

File: lab_5_scrapper/scrapper.py [hunk 23]
```diff
@@ -204,13 +416,40 @@ def prepare_environment(base_path: Union[pathlib.Path, str]) -> None:
     Args:
         base_path (Union[pathlib.Path, str]): Path where articles stores
     """
+    if not base_path.is_dir():
+        base_path.mkdir(parents=True, exist_ok=True)
+    elif base_path.exists():
+        shutil.rmtree(base_path)
+        base_path.mkdir(parents=True, exist_ok=True)
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py:423]: it is enough to remove directory first and then create it from scratch

File: lab_5_scrapper/scrapper.py [hunk 24]
```diff
@@ -204,13 +416,40 @@ def prepare_environment(base_path: Union[pathlib.Path, str]) -> None:
...
+    configuration = Config(constants.CRAWLER_CONFIG_PATH)
+
+    base_path = constants.ASSETS_PATH
+    prepare_environment(base_path)
+
+    crawler = Crawler(configuration)
+
+    crawler.find_articles()
+
+    all_urls = crawler.urls
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py:440]: unnecessary variable

File: lab_5_scrapper/scrapper.py [hunk 25]
```diff
@@ -204,13 +416,40 @@ def prepare_environment(base_path: Union[pathlib.Path, str]) -> None:
...
+    for index, url in enumerate(all_urls):
+        article_id = index + 1
+        html_parser = HTMLParser(full_url=url, article_id=article_id, config=configuration)
+        article = html_parser.parse()
+
+        if isinstance(article, Article):
+            to_raw(article)
+            to_meta(article)
+
+    print('Ready!')
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py:451]: remove debug print

File: lab_5_scrapper/scrapper.py [hunk 26]
```diff
@@ -168,6 +345,14 @@ def _fill_article_with_text(self, article_soup: BeautifulSoup) -> None:
         Args:
             article_soup (bs4.BeautifulSoup): BeautifulSoup instance
         """
+        main_div = article_soup.find('div', class_='article__content clearfix')
+        texts = []
+        if main_div:
+            all_ps = main_div.find_all('p', class_='article__text')
+            for p in all_ps:
+                texts.append(p.text)
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py:353]: replace with comprehension

**Review by demid5111 — CHANGES_REQUESTED** [review]
for 8 it is more or less ok, need to polish minor things

File: lab_5_scrapper/scrapper.py (hunk 9 above)
Comment (velgarita) [reply lab_5_scrapper/scrapper.py:100]: Done

File: lab_5_scrapper/scrapper.py (hunk 26 above)
Comment (velgarita) [reply lab_5_scrapper/scrapper.py:353]: Changed it

File: lab_5_scrapper/scrapper.py (hunk 22 above)
Comment (velgarita) [reply lab_5_scrapper/scrapper.py:405]: Fixed it

File: lab_5_scrapper/scrapper.py (hunk 23 above)
Comment (velgarita) [reply lab_5_scrapper/scrapper.py:423]: Changed it

File: lab_5_scrapper/scrapper.py (hunk 24 above)
Comment (velgarita) [reply lab_5_scrapper/scrapper.py:440]: Fixed it

File: lab_5_scrapper/scrapper.py (hunk 25 above)
Comment (velgarita) [reply lab_5_scrapper/scrapper.py:451]: Done

File: requirements.txt (hunk 15 above)
[the remark this answers is not in the document]
Comment (velgarita) [reply requirements.txt:2]: Sorry, now I removed it

File: requirements.txt (hunk 21 above)
Comment (velgarita) [reply requirements.txt:3]: Done

File: lab_5_scrapper/scrapper.py [hunk 27]
```diff
@@ -127,11 +256,45 @@ def _extract_url(self, article_bs: BeautifulSoup) -> str:
...
+        seed_urls = self.get_search_urls()
+
+        while len(self.urls) < self.config.get_num_articles():
+
+            for seed_url in seed_urls:
+                response = make_request(seed_url, self.config)
+                if not response.ok:
+                    continue
+
+                article_soup = BeautifulSoup(response.text, features='lxml')
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py:285]: .

File: lab_5_scrapper/scrapper.py [hunk 28]
```diff
@@ -26,11 +95,54 @@ def _extract_config_content(self) -> ConfigDTO:
...
+            ):
+            raise IncorrectHeadersError
+
+        if not isinstance(config.encoding, str):
+            raise IncorrectEncodingError
+
+        if not(
+            isinstance(config.timeout, int)
+            and 0 <= config.timeout <= 60
+        ):
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py:137]: .

File: lab_5_scrapper/scrapper.py [hunk 29]
```diff
@@ -26,11 +95,54 @@ def _extract_config_content(self) -> ConfigDTO:
...
+            raise IncorrectTimeoutError
+
+        if not (
+            isinstance(config.should_verify_certificate, bool)
+            and isinstance(config.headless_mode, bool)
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py:142]: .

File: lab_5_scrapper/scrapper.py [hunk 30]
```diff
@@ -26,11 +95,54 @@ def _extract_config_content(self) -> ConfigDTO:
         Returns:
             ConfigDTO: Config values
         """
+        with open(file=self.path_to_config, mode='r', encoding='utf-8') as file:
+            config_content = json.load(file)
+        return ConfigDTO(**{key: config_content[key] for key in config_content.keys()})
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py:100]: it is already a dict

**Review by demid5111 — CHANGES_REQUESTED** [review]

**Pull request closed without merging**